File: RemoteBees-iOS/RemoteBeesService/Couchbase/CouchbaseLite.swift
//
//  CouchbaseLite.swift
//  RemoteBees
//

import CouchbaseLiteSwift

class CouchbaseLite {
    
    private struct PreferencesDocument: Codable {
        let localeIdentifier: String
    }
    
    let database: Database
    private (set) var isClosed: Bool = false
    var isOpen: Bool {
        return !self.isClosed
    }

    init(name: String, localeIdentifier: String) throws {
        self.database = try Database(name: name)

        let prefs: PreferencesDocument = try {
            if let savedPrefs = try? self.getPreferences() {
                return savedPrefs
            }
            let newPrefs = PreferencesDocument(localeIdentifier: localeIdentifier)
            try self.savePreferences(newPrefs)
            return newPrefs
        }()

        if (localeIdentifier != prefs.localeIdentifier) {
            try self.clearUserDocuments()
            try self.savePreferences(PreferencesDocument(localeIdentifier: localeIdentifier))
        }
    }

    func clearUserDocuments() throws {
        guard self.isOpen else { return }
        try self.database.inBatch {
//            if let prefsDoc = self.database.document(withID: "<Document_Name>") {
//                try self.database.deleteDocument(prefsDoc)
//            }
            // Add more later
        }
    }

    func close() throws {
        try self.database.close()
        self.isClosed = true
    }

    private func savePreferences(_ prefs: PreferencesDocument) throws {
        guard self.isOpen else { return }
        let data = try DictionaryEncoder().encode(prefs)
        let document = MutableDocument(id: "PreferencesDocument", data: data)
        try self.database.saveDocument(document)
    }

    private func getPreferences() throws -> PreferencesDocument? {
        guard self.isOpen else { return nil }
        guard let document = self.database.document(withID: "PreferencesDocument") else { return nil }
        let dict = document.toDictionary()
        return try DictionaryDecoder().decode(PreferencesDocument.self, from: dict)
    }

}

File: RemoteBees-Android/RemoteBeesService/src/main/java/dev/beehive/remotebees/service/internal/couchbase/CouchbaseLite.kt
/***************************************************************************
 * This source file is part of the RemoteBees open source project.         *
 *                                                                         *
 * Licensed under the MIT License. See LICENSE for license information     *
 ***************************************************************************/

package dev.beehive.remotebees.service.internal.couchbase

import android.content.Context
import com.couchbase.lite.CouchbaseLite
import com.couchbase.lite.Database
import com.couchbase.lite.DatabaseConfiguration
import com.couchbase.lite.MutableDocument

class CouchbaseLiteDb(
    private val context: Context
    , private val name: String
    , private val localeIdentifier: String
) {
    private data class PreferencesDocument(
        val localeIdentifier: String
    )

    val database: Database

    init {
        CouchbaseLite.init(this.context)
        this.database = Database(name, DatabaseConfiguration())

        val prefs: PreferencesDocument = this.getPreferences() ?: {
            val newPrefs = PreferencesDocument(localeIdentifier = this.localeIdentifier)
                this.savePreferences(newPrefs)
                newPrefs
        }()

        if (this.localeIdentifier != prefs.localeIdentifier) {
            this.clearUserDocuments()
            this.savePreferences(PreferencesDocument(localeIdentifier = localeIdentifier))
        }
    }

    @Throws
    fun clearUserDocuments() {
        this.database.inBatch {
    //        this.getDocument("<DocumentName>")?.let(this::delete)
            // Add more later
        }
    }

    @Throws
    private fun savePreferences(prefs: PreferencesDocument) {
        val data = DictionaryEncoder().encode(prefs) ?: return
        val document = MutableDocument("PreferencesDocument", data)
        this.database.save(document)
    }

    @Throws
    private fun getPreferences(): PreferencesDocument? {
        val document = this.database.getDocument("PreferencesDocument") ?: return null
        val data = document.toMap()
        return DictionaryDecoder().decode(type = PreferencesDocument::class.java, value = data)
    }

}
